src/data/finsupp/basic.lean

```diff
@@ -482,80 +486,71 @@ end erase
   f.prod g = ∏ i, g i (f i) :=
 f.prod_of_support_subset (subset_univ _) g (λ x _, h x)
 
 @[simp, to_additive]
-lemma prod_single_index [has_zero β] [comm_monoid γ] {a : α} {b : β} {h : α → β → γ}
-  (h_zero : h a 0 = 1) :
+lemma prod_single_index {a : α} {b : M} {h : α → M → N} (h_zero : h a 0 = 1) :
   (single a b).prod h = h a b :=
 calc (single a b).prod h = ∏ x in {a}, h x (single a b x) :
   prod_of_support_subset _ support_single_subset h $ λ x hx, (mem_singleton.1 hx).symm ▸ h_zero
 ... = h a b : by simp
 
 @[to_additive]
-lemma prod_map_range_index [has_zero β₁] [has_zero β₂] [comm_monoid γ]
-  {f : β₁ → β₂} {hf : f 0 = 0} {g : α →₀ β₁} {h : α → β₂ → γ} (h0 : ∀a, h a 0 = 1) :
-  (map_range f hf g).prod h = g.prod (λa b, h a (f b)) :=
+lemma prod_map_range_index {f : M → M'} {hf : f 0 = 0} {g : α →₀ M} {h : α → M' → N}
+  (h0 : ∀a, h a 0 = 1) : (map_range f hf g).prod h = g.prod (λa b, h a (f b)) :=
 finset.prod_subset support_map_range $ λ _ _ H,
   by rw [not_mem_support_iff.1 H, h0]
 
-@[to_additive]
-lemma prod_zero_index [add_comm_monoid β] [comm_monoid γ] {h : α → β → γ} :
-  (0 : α →₀ β).prod h = 1 :=
-rfl
+@[simp, to_additive]
+lemma prod_zero_index {h : α → M → N} : (0 : α →₀ M).prod h = 1 := rfl
 
 @[to_additive]
-lemma prod_comm {α' : Type*} [has_zero β] {β' : Type*} [has_zero β'] (f : α →₀ β) (g : α' →₀ β')
-  [comm_monoid γ] (h : α → β → α' → β' → γ) :
+lemma prod_comm (f : α →₀ M) (g : β →₀ M') (h : α → M → β → M' → N) :
   f.prod (λ x v, g.prod (λ x' v', h x v x' v')) = g.prod (λ x' v', f.prod (λ x v, h x v x' v')) :=
-begin
-  dsimp [finsupp.prod],
-  rw finset.prod_comm,
-end
+finset.prod_comm
 
 @[simp, to_additive]
-lemma prod_ite_eq [has_zero β] [comm_monoid γ] (f : α →₀ β) (a : α) (b : α → β → γ) :
+lemma prod_ite_eq (f : α →₀ M) (a : α) (b : α → M → N) :
   f.prod (λ x v, ite (a = x) (b x v) 1) = ite (a ∈ f.support) (b a (f a)) 1 :=
 by { dsimp [finsupp.prod], rw f.support.prod_ite_eq, }
 
 /-- A restatement of `prod_ite_eq` with the equality test reversed. -/
 @[simp, to_additive "A restatement of `sum_ite_eq` with the equality test reversed."]
-lemma prod_ite_eq' [has_zero β] [comm_monoid γ] (f : α →₀ β) (a : α) (b : α → β → γ) :
+lemma prod_ite_eq' (f : α →₀ M) (a : α) (b : α → M → N) :
   f.prod (λ x v, ite (x = a) (b x v) 1) = ite (a ∈ f.support) (b a (f a)) 1 :=
 by { dsimp [finsupp.prod], rw f.support.prod_ite_eq', }
 
-@[simp] lemma prod_pow [fintype α] [comm_monoid γ] (f : α →₀ ℕ) (g : α → γ) :
+@[simp] lemma prod_pow [fintype α] (f : α →₀ ℕ) (g : α → N) :
   f.prod (λ a b, g a ^ b) = ∏ a, g a ^ (f a) :=
-begin
-  apply prod_subset (finset.subset_univ _),
-  intros a _ ha,
-  simp only [finsupp.not_mem_support_iff.mp ha, pow_zero]
-end
+f.prod_fintype _ $ λ a, pow_zero _
 
-/-- If `g` maps a second argument of 0 to 0, summing it over the
-result of `on_finset` is the same as summing it over the original
+/-- If `g` maps a second argument of 0 to 1, then multiplying it over the
```

Comment:
These changes to the docstring don't look like what the type says.

Reply:
I was cherry-picking from a branch where I turn it into `on_finset_prod` with `to_additive`. I'll PR this branch later.